ssl/quic/quic_dummy_handshake.c

```diff
@@ -0,0 +1,343 @@
+#define QUIC_DHS_STATE_RECEIVED_CERT        4
+#define QUIC_DHS_STATE_RECEIVED_CERT_VERIFY 5
+#define QUIC_DHS_STATE_RECEIVED_FINISHED    6
+#define QUIC_DHS_STATE_SENT_FINISHED        7
+
+#define QUIC_DHS_STATE_ERROR        0xFF
+
+struct quic_dhs_st {
+    QUIC_DHS_ARGS   args;
+    unsigned char   state;
+    unsigned char   *remote_transport_params;
+    size_t          remote_transport_params_len;
+    const unsigned char *local_transport_params;
+    size_t          local_transport_params_len;
+    unsigned char   rx_hdr[4];
+    size_t          rx_hdr_bytes_read;
+    size_t          rx_ee_bytes_read;
+};
+
+QUIC_DHS *ossl_quic_dhs_new(const QUIC_DHS_ARGS *args)
+{
+    QUIC_DHS *dhs;
+
+    if (args->crypto_send_cb == NULL
+        || args->crypto_recv_cb == NULL)
+        return NULL;
+
+    dhs = OPENSSL_zalloc(sizeof(*dhs));
+    if (dhs == NULL)
+        return NULL;
+
+    dhs->args   = *args;
+    dhs->state  = QUIC_DHS_STATE_INITIAL;
+    return dhs;
+}
+
+void ossl_quic_dhs_free(QUIC_DHS *dhs)
+{
+    if (dhs == NULL)
+        return;
+
+    OPENSSL_free(dhs->remote_transport_params);
+    OPENSSL_free(dhs);
+}
+
+static int dhs_send(QUIC_DHS *dhs, unsigned char type,
+                    const void *buf, size_t buf_len)
+{
+    size_t consumed = 0;
+    uint32_t len;
+    unsigned char hdr[4];
+
+    len = buf_len;
+    hdr[0] = type;
+    hdr[1] = (len >> 16) & 0xFF;
+    hdr[2] = (len >>  8) & 0xFF;
+    hdr[3] = (len      ) & 0xFF;
+
+    if (!dhs->args.crypto_send_cb(hdr, sizeof(hdr), &consumed,
+                                  dhs->args.crypto_send_cb_arg)
+        || consumed < sizeof(hdr)
+        || (buf_len > 0 && (!dhs->args.crypto_send_cb(buf, buf_len, &consumed,
+                                                      dhs->args.crypto_send_cb_arg)
+                            || consumed < buf_len)))
+        /*
+         * We do not handle a full buffer here properly but the DHS produces so
+         * little data this should not matter. By the time we want to fix this
+         * the real handshake layer will be ready.
+         */
+        return 0;
+
+    return 1;
+}
+
+static int dhs_recv_sof(QUIC_DHS *dhs, uint32_t *type, size_t *frame_len)
+{
+    size_t bytes_read = 0;
+    uint32_t l;
+
+    if (!dhs->args.crypto_recv_cb(dhs->rx_hdr + dhs->rx_hdr_bytes_read,
+                                  sizeof(dhs->rx_hdr) - dhs->rx_hdr_bytes_read,
+                                  &bytes_read,
+                                  dhs->args.crypto_recv_cb_arg))
+        return 0;
+
+    dhs->rx_hdr_bytes_read += bytes_read;
+    if (dhs->rx_hdr_bytes_read < sizeof(dhs->rx_hdr)) {
+        /* Not got entire header yet. */
+        *type       = UINT32_MAX;
+        *frame_len  = 0;
+        return 2;
+    }
+
+    l = (((uint32_t)dhs->rx_hdr[1]) << 16)
+      | (((uint32_t)dhs->rx_hdr[2]) <<  8)
+      |   (uint32_t)dhs->rx_hdr[3];
+
+    if (l > SIZE_MAX)
+        return 0;
```

Comment:
ssl/quic/quic_dummy_handshake.c:123:11: warning: result of comparison of constant 18446744073709551615 with expression of type 'uint32_t' (aka 'unsigned int') is always false [-Wtautological-constant-out-of-range-compare]
    if (l > SIZE_MAX)
        ~ ^ ~~~~~~~~

Reply:
This warning is supposed to be disabled in `Configure`. I consider this warning harmful since it is actively encouraging people to write less correct code.

Reply:
Added some thoughts on this warning below

Reply:
It's unclear to me why I still get the warning.